browser/ui/views/brave_shields/cookie_consent_browser_helper.cc

```diff
@@ -0,0 +1,218 @@
+ public:
+  static inline constexpr char kKey[] =
+      "BraveCookieConsentBlockerProfileHelperData";
+
+  CookieConsentProfileHelper() {}
+  ~CookieConsentProfileHelper() override = default;
+
+  static CookieConsentProfileHelper* GetOrCreateForProfile(Profile* profile) {
+    DCHECK(profile);
+    auto* helper =
+        static_cast<CookieConsentProfileHelper*>(profile->GetUserData(kKey));
+    if (!helper) {
+      auto instance = std::make_unique<CookieConsentProfileHelper>();
+      helper = instance.get();
+      profile->SetUserData(kKey, std::move(instance));
+    }
+    return helper;
+  }
+
+  static void RemoveFromProfile(Profile* profile) {
+    DCHECK(profile);
+    profile->RemoveUserData(kKey);
+  }
+
+  base::flat_set<std::string>& host_set() { return host_set_; }
+  const base::flat_set<std::string>& host_set() const { return host_set_; }
+
+ private:
+  base::flat_set<std::string> host_set_;
+};
+
+bool ShouldEventuallyShowBubble() {
+  auto* local_state = g_browser_process->local_state();
+  if (local_state->GetBoolean(prefs::kAdBlockCookieConsentBlockerDisplayed)) {
+    return false;
+  }
+
+  auto* regional_service_manager =
+      g_brave_browser_process->ad_block_service()->regional_service_manager();
+  DCHECK(regional_service_manager);
+  if (regional_service_manager->IsFilterListEnabled(kCookieListUuid)) {
+    return false;
+  }
+
+  return true;
+}
+
+void EventuallyRemoveFromBrowser(
+    base::WeakPtr<CookieConsentBrowserHelper> instance) {
+  DCHECK(instance);
+
+  auto callback = [](base::WeakPtr<CookieConsentBrowserHelper> instance) {
+    if (instance) {
+      CookieConsentBrowserHelper::RemoveFromBrowser(&instance->GetBrowser());
+    }
+  };
+
+  base::SequencedTaskRunnerHandle::Get()->PostTask(
+      FROM_HERE, base::BindOnce(callback, instance));
+}
+
+}  // namespace
+
+CookieConsentBrowserHelper::CookieConsentBrowserHelper(Browser* browser)
+    : BrowserUserData<CookieConsentBrowserHelper>(*browser) {
+  GetBrowser().tab_strip_model()->AddObserver(this);
+}
+
+CookieConsentBrowserHelper::~CookieConsentBrowserHelper() {
+  if (!ShouldEventuallyShowBubble()) {
+    CookieConsentProfileHelper::RemoveFromProfile(GetBrowser().profile());
+  }
+}
+
+void CookieConsentBrowserHelper::CreateForBrowserIfNecessary(Browser* browser) {
+  if (ShouldEventuallyShowBubble() &&
+      base::FeatureList::IsEnabled(features::kBraveAdblockCookieListBubble)) {
+    CreateForBrowser(browser);
+  }
+}
+
+bool CookieConsentBrowserHelper::ShowBubble() {
+  // Do not show the bubble if this is not the currently active browser window.
+  auto* browser_view = BrowserView::GetBrowserViewForBrowser(&GetBrowser());
+  if (!browser_view || !browser_view->IsActive()) {
+    return false;
+  }
+
+  // Do not show the bubble if the filter list is not yet available, likely
+  // because the regional filter list has not yet been donwloaded.
+  auto* regional_service_manager =
```

Comment:
same thing here, pass in the regional service manager to the constructor to minimize dependencies